File: python3/l0039_combination_sum.py
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        def combinationSumImpl(candidates: List[int], target: int) -> List[List[int]]:
            result = []
            for i, c in enumerate(candidates):
                if c == target:
                    result.append([c])
                elif c < target:
                    for r in combinationSumImpl(candidates[i:], target - c):
                        result.append([c] + r)
            return result
        
        return combinationSumImpl(candidates, target)

File: java/src/main/java/leetcode/L0039CombinationSum.java
package leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class L0039CombinationSum {
    private List<List<Integer>> combination(int[] candidates, int target, int start) {
        if (target == 0) {
            return Arrays.asList(new ArrayList<Integer>());
        }
        List<List<Integer>> result = new ArrayList<>();
        for (int i = start; i < candidates.length; i++) {
            int candidate = candidates[i];
            if (candidate > target) {
                continue;
            }
            for (List<Integer> r : combination(candidates, target-candidate, i)) {
                List<Integer> comb = new ArrayList<Integer>();
                comb.add(candidate);
                comb.addAll(r);
                result.add(comb);
            }
        }
        return result;
    }
  
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        return combination(candidates, target, 0);  
    }
}
